File: problems/802_find-eventual-safe-states/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
    bool dfs(int u, vector<vector<int>> &graph, vector<char> &state)
    {
        if (state[u] == 1)
            return false; // found a cycle => unsafe
        if (state[u] == 2)
            return true; // already known safe

        state[u] = 1; // mark visiting
        for (int v : graph[u])
        {
            if (!dfs(v, graph, state))
                return false;
        }
        state[u] = 2; // all neighbors safe => safe
        return true;
    }

    vector<int> eventualSafeNodes(vector<vector<int>> &graph)
    {
        int n = (int)graph.size();
        vector<char> state(n, 0);
        vector<int> safe;
        safe.reserve(n);

        for (int i = 0; i < n; ++i)
        {
            if (dfs(i, graph, state))
                safe.push_back(i);
        }
        return safe; // already in ascending order
    }
    bool dfsDetectCycle(int node, vector<vector<int>> &graph, vector<bool> &vis, vector<bool> &path, vector<int> &safenodes)
    {
        vis[node] = true;
        path[node] = true;
        for (auto nei : graph[node])
        {
            if (!vis[nei])
            {
                if (dfsDetectCycle(nei, graph, vis, path, safenodes))
                    return true;
            }
            else if (path[nei])
                return true;
        }
        safenodes.push_back(node);
        path[node] = false;
        return false;
    }
    vector<int> eventualSafeNodes_2(vector<vector<int>> &graph)
    {
        vector<bool> vis(graph.size(), false), path(graph.size(), false);
        vector<int> safenodes;
        safenodes.reserve(graph.size());
        for (int i = 0; i < graph.size(); i++)
        {
            if (!vis[i])
                dfsDetectCycle(i, graph, vis, path, safenodes);
        }
        sort(safenodes.begin(), safenodes.end());
        return safenodes;
    }
    vector<int> eventualSafeNodes_3(vector<vector<int>> &graph)
    {
        int n = graph.size();
        vector<vector<int>> rev(n);
        vector<int> outdegree(n, 0);
        for (int i = 0; i < n; i++)
        {
            outdegree[i] = graph[i].size();
            for (auto nei : graph[i])
            {
                rev[nei].push_back(i);
            }
        }
        queue<int> q;
        for (int i = 0; i < n; i++)
            if (outdegree[i] == 0)
                q.push(i);

        vector<int> safenodes;
        safenodes.reserve(n);
        while (!q.empty())
        {
            int curr = q.front();
            q.pop();
            safenodes.push_back(curr);
            for (auto nei : rev[curr])
            {
                outdegree[nei]--;
                if (outdegree[nei] == 0)
                    q.push(nei);
            }
        }
        sort(safenodes.begin(), safenodes.end());
        return safenodes;
    }
};

void printVector(const vector<int> &values)
{
    cout << "[";
    for (size_t i = 0; i < values.size(); ++i)
    {
        cout << values[i];
        if (i + 1 < values.size())
            cout << ", ";
    }
    cout << "]";
}

int main()
{
    Solution sol;

    vector<vector<int>> sample1 = {{1, 2}, {2, 3}, {5}, {0}, {5}, {}, {}};
    vector<vector<int>> sample2 = {{1, 2, 3, 4}, {1, 2}, {3, 4}, {0, 4}, {}};

    auto result1 = sol.eventualSafeNodes(sample1);
    cout << "Sample 1 placeholder result: ";
    printVector(result1);
    cout << "\nSample 1 expected output: [2,4,5,6]\n\n";

    auto result2 = sol.eventualSafeNodes(sample2);
    cout << "Sample 2 placeholder result: ";
    printVector(result2);
    cout << "\nSample 2 expected output: [4]\n";

    return 0;
}

File: problems/802_find-eventual-safe-states/main.rs
struct Solution;

impl Solution {
    pub fn eventual_safe_nodes(graph: Vec<Vec<i32>>) -> Vec<i32> {
        let n = graph.len();
        let mut state = vec![0u8; n]; // 0: unvisited, 1: visiting, 2: safe
        let mut safe = Vec::new();
        for i in 0..n {
            if Self::dfs(i, &graph, &mut state) {
                safe.push(i as i32);
            }
        }
        safe
    }

    fn dfs(u: usize, graph: &Vec<Vec<i32>>, state: &mut Vec<u8>) -> bool {
        if state[u] == 1 {
            return false; // cycle
        }
        if state[u] == 2 {
            return true; // safe
        }
        state[u] = 1; // visiting
        for &v in &graph[u] {
            let v = v as usize;
            if !Self::dfs(v, graph, state) {
                return false;
            }
        }
        state[u] = 2; // safe
        true
    }

    fn eventual_safe_nodes_2(graph: Vec<Vec<i32>>) -> Vec<i32> {
        let n = graph.len();
        let mut vis = vec![false; n];
        let mut path = vec![false; n];
        let mut safenodes = Vec::new();
        for i in 0..n {
            if !vis[i] {
                Self::dfs_detect_cycle(i, &graph, &mut vis, &mut path, &mut safenodes);
            }
        }
        safenodes.sort();
        safenodes
    }

    fn dfs_detect_cycle(
        node: usize,
        graph: &Vec<Vec<i32>>,
        vis: &mut Vec<bool>,
        path: &mut Vec<bool>,
        safenodes: &mut Vec<i32>,
    ) -> bool {
        vis[node] = true;
        path[node] = true;
        for &nei in &graph[node] {
            let nei = nei as usize;
            if !vis[nei] {
                if Self::dfs_detect_cycle(nei, graph, vis, path, safenodes) {
                    return true;
                }
            } else if path[nei] {
                return true;
            }
        }
        safenodes.push(node as i32);
        path[node] = false;
        false
    }

    fn eventual_safe_nodes_3(graph: Vec<Vec<i32>>) -> Vec<i32> {
        let n = graph.len();
        let mut rev: Vec<Vec<usize>> = vec![vec![]; n];
        let mut outdegree: Vec<i32> = vec![0; n];
        for i in 0..n {
            outdegree[i] = graph[i].len() as i32;
            for &nei in &graph[i] {
                let nei = nei as usize;
                rev[nei].push(i);
            }
        }
        use std::collections::VecDeque;
        let mut q = VecDeque::new();
        for i in 0..n {
            if outdegree[i] == 0 {
                q.push_back(i);
            }
        }
        let mut safenodes = Vec::new();
        while let Some(curr) = q.pop_front() {
            safenodes.push(curr as i32);
            for &nei in &rev[curr] {
                outdegree[nei] -= 1;
                if outdegree[nei] == 0 {
                    q.push_back(nei);
                }
            }
        }
        safenodes.sort();
        safenodes
    }
}

fn main() {
    let sample1 = vec![
        vec![1, 2],
        vec![2, 3],
        vec![5],
        vec![0],
        vec![5],
        vec![],
        vec![],
    ];
    let sample2 = vec![vec![1, 2, 3, 4], vec![1, 2], vec![3, 4], vec![0, 4], vec![]];

    let result1 = Solution::eventual_safe_nodes(sample1);
    println!("Sample 1 placeholder result: {:?}", result1);
    println!("Sample 1 expected output: [2,4,5,6]\n");

    let result2 = Solution::eventual_safe_nodes(sample2);
    println!("Sample 2 placeholder result: {:?}", result2);
    println!("Sample 2 expected output: [4]");
}
